src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetComInterfaceForObjectTests.cs

```diff
@@ -0,0 +1,231 @@
+            }
+            finally
+            {
+                Marshal.Release(iUnknown);
+            }
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void GetComInterfaceForObject_GenericWithValidStruct_ReturnsExpected()
+        {
+            var o = new StructWithInterface();
+            IntPtr iUnknown = Marshal.GetComInterfaceForObject<StructWithInterface, NonGenericInterface>(o);
+            try
+            {
+                Assert.NotEqual(IntPtr.Zero, iUnknown);
+            }
+            finally
+            {
+                Marshal.Release(iUnknown);
+            }
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void GetComInterfaceForObject_NonGenericWithValidClass_ReturnsExpected()
+        {
+            var o = new ClassWithInterface();
+            IntPtr iUnknown = Marshal.GetComInterfaceForObject(o, typeof(NonGenericInterface));
+            try
+            {
+                Assert.NotEqual(IntPtr.Zero, iUnknown);
+            }
+            finally
+            {
+                Marshal.Release(iUnknown);
+            }
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void GetComInterfaceForObject_NonGenericWithValidStruct_ReturnsExpected()
+        {
+            var o = new StructWithInterface();
+            IntPtr iUnknown = Marshal.GetComInterfaceForObject(o, typeof(NonGenericInterface));
+            try
+            {
+                Assert.NotEqual(IntPtr.Zero, iUnknown);
+            }
+            finally
+            {
+                Marshal.Release(iUnknown);
+            }
+        }
+
+        [Theory]
+        [InlineData(CustomQueryInterfaceMode.Allow)]
+        [InlineData(CustomQueryInterfaceMode.Ignore)]
+        [InlineData(CustomQueryInterfaceMode.Allow + 1)]
+        [InlineData(CustomQueryInterfaceMode.Ignore - 1)]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void GetComInterfaceForObject_NonGenericCustomQueryInterfaceModeWithValidClass_ReturnsExpected(CustomQueryInterfaceMode mode)
+        {
+            var o = new ClassWithInterface();
+            IntPtr iUnknown = Marshal.GetComInterfaceForObject(o, typeof(NonGenericInterface));
+            try
+            {
+                Assert.NotEqual(IntPtr.Zero, iUnknown);
+            }
+            finally
+            {
+                Marshal.Release(iUnknown);
+            }
+        }
+
+        [Theory]
+        [InlineData(CustomQueryInterfaceMode.Allow)]
+        [InlineData(CustomQueryInterfaceMode.Ignore)]
+        [InlineData(CustomQueryInterfaceMode.Allow + 1)]
+        [InlineData(CustomQueryInterfaceMode.Ignore - 1)]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void GetComInterfaceForObject_NonGenericCustomQueryInterfaceModeWithValidStruct_ReturnsExpected(CustomQueryInterfaceMode mode)
```

Comment:
Please add a case: managedclass implements ICustomQueryInterface to better testing CustomQueryInterfaceMode

Reply:
Sure - I just submitted a commit with PR feedback but can't really add these tests now as I'm on my mac. Either I can submit a follow up PR or add these tests in the next week

I also think I might've found an incorrect assertion inside the implementation for `ICustomQueryInterface` usage:

https://github.com/dotnet/coreclr/blob/master/src/vm/comcallablewrapper.cpp#L3358

```cpp
bool ComCallWrapper::GetComIPFromCCW_HandleCustomQI(
                ComCallWrapper * pWrap, REFIID riid, MethodTable * pIntfMT, IUnknown ** ppUnkOut)
{
    CONTRACTL
    {
        THROWS;
        GC_TRIGGERS;
        MODE_ANY;
        PRECONDITION(CheckPointer(pWrap));
    }
    CONTRACTL_END;

    // Customize QI: We call the method System.Runtime.InteropServices.ICustomQueryInterface
    //               GetInterface implemented by user to do the customized QI work.
    CustomQueryInterfaceResult retVal = Handled;

    // prepare the GUID
    GUID guid;
    if (IsEqualGUID(riid, GUID_NULL) && pIntfMT != NULL)
    {
        // riid is null, we retrieve the guid from the methodtable
        pIntfMT->GetGuid(&guid, true);
    }
    else
    {
        // copy riid to avoid user modify it
        guid = riid;
    }

    InvokeICustomQueryInterfaceGetInterfaceArgs args = {pWrap, &guid, ppUnkOut, &retVal};

    ADID targetADID;
    Context *pTargetContext;
    if (pWrap->NeedToSwitchDomains(GetThread(), &targetADID, &pTargetContext))
        InvokeICustomQueryInterfaceGetInterface_AppDomainTransition(&args, targetADID, pTargetContext);
    else
        InvokeICustomQueryInterfaceGetInterface_CallBack(&args);
    // return if user already handle the QI
    if (retVal == Handled)
        return true;
    // return NULL if user wants to fail the QI
    if (retVal == Failed)
    {
        *ppUnkOut = NULL;
        return true;
    }
    // assure that user returns the known return value
    _ASSERTE(retVal == NotHandled);
    return false;
}
```

We are free to return any value from the method
```
public System.Runtime.InteropServices.CustomQueryInterfaceResult GetInterface (ref Guid iid, out IntPtr ppv);
```

This means that we can trigger the assertion if we return an unknown value, right?

Reply:
Also, any idea why `GetComInterfaceForObjectInContext` is not exposed in .NET Core? It would be relatively simple to implement as the `GetComInterfaceForObjectNative` takes a `CLR_BOOL fOnlyInContext` parameter already...

If we don't want to expose it, I can submit a PR removing it from the code

Reply:
>We are free to return any value from the method

@hughbe How is that possible? It is typed as an enum and there are only 3 values. What am I missing here?

Reply:
The following code is legal in c#
```
public class Custom : ICustomQueryInterface
	{
		public CustomQueryInterfaceResult GetInterface(ref Guid iid, out IntPtr ppv)
		{
			ppv = IntPtr.Zero;
			return CustomQueryInterfaceResult.Failed + 1;
		}
	}
```

Reply:
Okay. This is legal in the sense that a) it will compile and b) that it will be typed. I agree. However it is not legal in the API sense to return any arbitrary value and pass it off as a valid enum value. The assert is correct because there are only 3 defined types and they are handled. If we follow the logic of the above we would need to validate the entire `Int32` space. The above code is legal C# but invalid usage of the API.